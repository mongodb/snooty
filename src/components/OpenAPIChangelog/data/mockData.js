export const mockChangelog = [
  {
    date: '2024-02-01',
    paths: [
      {
        path: '/api/atlas/v2/groups/{groupId}/alertConfigs/{alertConfigId}',
        httpMethod: 'PUT',
        operationId: 'updateAlertConfiguration',
        tag: 'Alert Configurations',
        versions: [
          {
            version: '2023-01-01',
            stabilityLevel: 'stable',
            changeType: 'removed',
            changes: [
              {
                change: 'resource version 2023-01-01 was removed, latest available resource version is 2023-02-01',
                changeCode: 'resource-version-removed',
                backwardCompatible: true,
              },
            ],
          },
        ],
      },
    ],
  },
  {
    date: '2023-02-01',
    paths: [
      {
        path: '/api/atlas/v2/groups/byName/{groupName}',
        httpMethod: 'GET',
        operationId: 'getProjectByName',
        tag: 'Projects',
        versions: [
          {
            version: '2023-02-01',
            stabilityLevel: 'stable',
            changeType: 'release',
            changes: [
              {
                change: 'resource version added.',
                changeCode: 'resource-version-added',
                backwardCompatible: false,
              },
            ],
          },
        ],
      },
      {
        path: '/api/atlas/v2/groups/{groupId}/alertConfigs/{alertConfigId}',
        httpMethod: 'PUT',
        operationId: 'updateAlertConfiguration',
        tag: 'Alert Configurations',
        versions: [
          {
            version: '2023-02-01',
            stabilityLevel: 'stable',
            changeType: 'release',
            changes: [
              {
                change: 'resource version added.',
                changeCode: 'resource-version-added',
                backwardCompatible: false,
              },
              {
                change:
                  "for the 'query' request parameter 'clusterName', default value was changed from 'default' to 'cluster0'",
                changeCode: 'request-parameter-default-value-changed',
                backwardCompatible: false,
              },
              {
                change: "added the optional property 'links' in the response with the '200' status.",
                changeCode: 'response-optional-property-added',
                backwardCompatible: true,
              },
            ],
          },
          {
            version: '2023-01-01',
            stabilityLevel: 'stable',
            changeType: 'deprecate',
            sunsetDate: '2024-02-01',
            changes: [
              {
                change: 'deprecated by resource version 2023-02-01 and marked for removal for 2024-02-01.',
                changeCode: 'resource-version-deprecated',
                backwardCompatible: false,
              },
            ],
          },
        ],
      },
      {
        path: '/api/atlas/v2/groups/{groupId}/accessList',
        httpMethod: 'POST',
        operationId: 'createProjectIpAccessList',
        tag: 'Project IP Access List',
        versions: [
          {
            version: '2023-01-01',
            stabilityLevel: 'stable',
            changeType: 'update',
            changes: [
              {
                change: "added the optional property 'name' in the response with the '200' status.",
                changeCode: 'response-optional-property-added',
                backwardCompatible: true,
              },
              {
                change: "added the optional property 'comment' in the response with the '200' status.",
                changeCode: 'response-optional-property-added',
                backwardCompatible: true,
              },
            ],
          },
        ],
      },
    ],
  },
  {
    date: '2023-02-22',
    paths: [
      {
        path: '/api/atlas/v2/groups/byName/{groupName}',
        httpMethod: 'GET',
        operationId: 'getProjectByName',
        tag: 'Projects',
        versions: [
          {
            version: '2023-02-01',
            stabilityLevel: 'stable',
            changeType: 'update',
            changes: [
              {
                change:
                  "added the optional property 'withDefaultAlertsSettings' in the response with the '200' status.",
                changeCode: 'response-optional-property-added',
                backwardCompatible: true,
              },
            ],
          },
        ],
      },
    ],
  },
];

export const mockDiff = [
  {
    path: '/api/atlas/v2/federationSettings/{federationSettingsId}/connectedOrgConfigs',
    httpMethod: 'GET',
    operationId: 'listConnectedOrgConfigs',
    tag: 'Federated Authentication',
    changeType: 'update',
    changes: [
      {
        change: "removed the non-success response with the status '400'.",
        changeCode: 'response-non-success-status-removed',
        backwardCompatible: false,
      },
      {
        change: "added the new optional 'query' request parameter 'envelope'.",
        changeCode: 'new-optional-request-parameter',
        backwardCompatible: true,
      },
    ],
  },
  {
    path: '/api/atlas/v2/groups/{groupId}/accessList',
    httpMethod: 'POST',
    operationId: 'createProjectIpAccessList',
    tag: 'Project IP Access List',
    changeType: 'removed',
    changes: [
      {
        change: "removed the required property 'name' from the response with the '200' status.",
        changeCode: 'response-required-property-removed',
        backwardCompatible: false,
      },
    ],
  },
];

export const mockIndex = {
  runDate: '2023-04-03',
  runId: '20230403105948111',
  specRevision: '422ab47f864909549362a7b39688404c82e2540b',
  specRevisionShort: '422ab47f864',
  versions: ['2023-01-01', '2023-02-01', '2023-08-11~preview'],
};
