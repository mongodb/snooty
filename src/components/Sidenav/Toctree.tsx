import React from 'react';
import { TocTreeEntry } from '../../types/ast';
import TOCNode from './TOCNode';

export type ToctreeProps = {
  toctree: TocTreeEntry;
  slug: string;
  handleClick: () => void;
};

const Toctree = ({ handleClick, slug, toctree: { children } }: ToctreeProps) => {
  return (
    <>
      {children.map((c, idx) => (
        <TOCNode key={idx} activeSection={slug} handleClick={handleClick} node={c} />
      ))}
    </>
  );
};

export default Toctree;
