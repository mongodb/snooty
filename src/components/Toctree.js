import React, { useCallback, useState } from 'react';
import PropTypes from 'prop-types';
import { css, cx } from '@leafygreen-ui/emotion';
import { SideNavItem } from '@leafygreen-ui/side-nav';
import Link from './Link';
import { formatText } from '../utils/format-text';

const isCurrentPage = (currentUrl, slug) => {
  const trimSlashes = (str) => str.replace(/^\/|\/$/g, '');
  if (!currentUrl || !slug) return false;
  return trimSlashes(currentUrl) === trimSlashes(slug);
};

const tocStyling = css`
  border-left: none;
  padding-top: 8px;
  padding-bottom: 8px;

  // TODO: Remove after mongodb-docs.css is removed
  :hover,
  :focus {
    color: unset;
  }
`;

const tocDrawer = css`
  background-color: unset !important;
  cursor: pointer !important;
`;

const Drawer = ({ children, ...props }) => {
  const [expand, setExpand] = useState(false);

  const handleClickExpand = useCallback(() => {
    setExpand((curr) => !curr);
  }, [setExpand]);

  return (
    <SideNavItem className={cx(tocStyling, tocDrawer)} active={expand} onClick={handleClickExpand} {...props}>
      {children}
    </SideNavItem>
  );
};

// LG's SideNavItem component only renders nested list items correctly if the component has a displayName
// of SideNavItem or SideNavGroup. This is kind of a hacky workaround since we want to use a SideNavItem as
// a drawer instead of a SideNavGroup.
Drawer.displayName = 'SideNavGroup';

// Use a recursive function instead of a recursive component since LG's SideNavItem is reliant on the
// "SideNavItem" displayName
const recurse = ({ currentSlug, node: { children, options, title, slug, url }, ...props }) => {
  const isActive = isCurrentPage(currentSlug, slug);
  const label = formatText(title);
  const target = slug || url;
  const childNodes = children.map((c) => recurse({ currentSlug, node: c, ...props }));

  if (children?.length > 0 && options?.drawer) {
    return (
      <Drawer key={`drawer-${target}`}>
        {label}
        {childNodes}
      </Drawer>
    );
  }

  return (
    <SideNavItem key={`node-${target}`} className={cx(tocStyling)} active={isActive} as={Link} to={target} {...props}>
      {label}
      {childNodes}
    </SideNavItem>
  );
};

const Toctree = ({ slug, toctree: { children } }) => {
  return <>{children.map((c) => recurse({ currentSlug: slug, node: c }))}</>;
};

Toctree.propTypes = {
  toctree: PropTypes.shape({
    children: PropTypes.arrayOf(
      PropTypes.shape({
        title: PropTypes.oneOfType([PropTypes.arrayOf(PropTypes.object), PropTypes.string]).isRequired,
        slug: PropTypes.string,
        url: PropTypes.string,
        children: PropTypes.array.isRequired,
        options: PropTypes.shape({
          drawer: PropTypes.bool,
          styles: PropTypes.objectOf(PropTypes.string),
        }),
      })
    ),
  }),
};

export default Toctree;
