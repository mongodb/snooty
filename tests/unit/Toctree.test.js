import React from 'react';
import { shallow } from 'enzyme';
import Toctree from '../../src/components/Toctree';
import mockData from './data/Toctree.test.json';

// mockData is a minimal toctree, based off of Realm's toctree:
// /get-started (drawer)
//   /get-started/introduction-mobile
//   /get-started/introduction-backend
// /sdk
//   /sdk/android
//     /sdk/android/fundamentals (drawer)
//       /sdk/android/fundamentals/async-api
//   /sdk/ios

const shallowToctree = (slug) => {
  return shallow(<Toctree slug={slug} toctree={mockData?.toctree} />);
};

describe('Toctree', () => {
  it('renders', () => {
    const wrapper = shallowToctree('/');
    expect(wrapper).toMatchSnapshot();
  });

  it('correctly assigns the active SideNavItem', () => {
    const checkForActiveSideNavItem = (wrapper) =>
      wrapper.findWhere((n) => n.name() === 'SideNavItem' && n.prop('active') === true);

    let testSlug = 'sdk/android/fundamentals/async-api';
    let wrapper = shallowToctree(`/${testSlug}/`);
    let activeSideNavItem = checkForActiveSideNavItem(wrapper);
    expect(activeSideNavItem).toHaveLength(1);
    expect(activeSideNavItem.prop('to')).toEqual(testSlug);

    testSlug = 'get-started/introduction-backend';
    wrapper = shallowToctree(testSlug);
    activeSideNavItem = checkForActiveSideNavItem(wrapper);
    expect(activeSideNavItem).toHaveLength(1);
    expect(activeSideNavItem.prop('to')).toEqual(testSlug);
  });
});
