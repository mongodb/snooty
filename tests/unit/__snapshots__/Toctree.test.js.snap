// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Toctree renders 1`] = `
.emotion-0 {
  border-left: none;
  padding-top: 8px;
  padding-bottom: 8px;
}

.emotion-0:hover,
.emotion-0:focus {
  color: unset;
}

<Fragment>
  <SideNavGroup
    key="drawer-get-started"
  >
    <ComponentFactory
      key="0"
      nodeData={
        Object {
          "position": Object {
            "start": Object {
              "line": 0,
            },
          },
          "type": "text",
          "value": "Get Started",
        }
      }
    />
    <SideNavItem
      active={false}
      as={[Function]}
      className="emotion-0"
      key="node-get-started/introduction-mobile"
      to="get-started/introduction-mobile"
    >
      <ComponentFactory
        key="0"
        nodeData={
          Object {
            "position": Object {
              "start": Object {
                "line": 0,
              },
            },
            "type": "text",
            "value": "Introduction for Mobile Developers",
          }
        }
      />
    </SideNavItem>
    <SideNavItem
      active={false}
      as={[Function]}
      className="emotion-0"
      key="node-get-started/introduction-backend"
      to="get-started/introduction-backend"
    >
      <ComponentFactory
        key="0"
        nodeData={
          Object {
            "position": Object {
              "start": Object {
                "line": 0,
              },
            },
            "type": "text",
            "value": "Introduction for Backend Developers",
          }
        }
      />
    </SideNavItem>
  </SideNavGroup>
  <SideNavItem
    active={false}
    as={[Function]}
    className="emotion-0"
    key="node-sdk"
    to="sdk"
  >
    <ComponentFactory
      key="0"
      nodeData={
        Object {
          "position": Object {
            "start": Object {
              "line": 0,
            },
          },
          "type": "text",
          "value": "Realm Database SDKs",
        }
      }
    />
    <SideNavItem
      active={false}
      as={[Function]}
      className="emotion-0"
      key="node-sdk/android"
      to="sdk/android"
    >
      <ComponentFactory
        key="0"
        nodeData={
          Object {
            "position": Object {
              "start": Object {
                "line": 0,
              },
            },
            "type": "text",
            "value": "Android SDK",
          }
        }
      />
      <SideNavGroup
        key="drawer-sdk/android/fundamentals"
      >
        <ComponentFactory
          key="0"
          nodeData={
            Object {
              "position": Object {
                "start": Object {
                  "line": 0,
                },
              },
              "type": "text",
              "value": "Fundamentals",
            }
          }
        />
        <SideNavItem
          active={false}
          as={[Function]}
          className="emotion-0"
          key="node-sdk/android/fundamentals/async-api"
          to="sdk/android/fundamentals/async-api"
        >
          <ComponentFactory
            key="0"
            nodeData={
              Object {
                "position": Object {
                  "start": Object {
                    "line": 0,
                  },
                },
                "type": "text",
                "value": "Asynchronous API",
              }
            }
          />
        </SideNavItem>
      </SideNavGroup>
    </SideNavItem>
    <SideNavItem
      active={false}
      as={[Function]}
      className="emotion-0"
      key="node-sdk/ios"
      to="sdk/ios"
    >
      <ComponentFactory
        key="0"
        nodeData={
          Object {
            "position": Object {
              "start": Object {
                "line": 0,
              },
            },
            "type": "text",
            "value": "iOS SDK",
          }
        }
      />
    </SideNavItem>
  </SideNavItem>
</Fragment>
`;
